File: elementary-graph-algorithms/python/depth_first_search.py
from graph import Color, Edge, Graph, Vertex

time = 0


def depth_first_search(graph):
    vertices = graph.vertices

    for u in vertices.values():
        u.color = Color.WHITE
        u.predecessor = None

    for u in vertices.values():
        if u.color == Color.WHITE:
            depth_first_search_internal(graph, u)


def depth_first_search_internal(graph, u):
    global time

    time += 1
    u.start_time = time
    u.color = Color.GRAY
    print(f"{u.name} change color to gray. time = {time}")

    adjacencies = graph.adjacency_list[u.name]
    for v in adjacencies:
        if v.color == Color.WHITE:
            v.predecessor = u
            depth_first_search_internal(graph, v)

    time += 1
    u.finish_time = time
    u.color = Color.BLACK
    print(f"{u.name} change color to black. time = {time}")


def print_path(u, u_name):
    if not u:
        return

    print_path(u.predecessor, u_name)
    print(u.name, end=" -> " if u.name != u_name else "")


if __name__ == "__main__":
    vertices = {}
    for name in "uvxywz":
        vertices[name] = Vertex(name)

    edges = []
    edges.append(Edge("u", "v"))
    edges.append(Edge("u", "x"))
    edges.append(Edge("v", "y"))
    edges.append(Edge("x", "v"))
    edges.append(Edge("y", "x"))
    edges.append(Edge("w", "y"))
    edges.append(Edge("w", "z"))
    edges.append(Edge("z", "z"))

    graph = Graph(vertices, edges)

    print("--- graph ---")
    graph.print_graph()

    print("--- depth first search ---")
    depth_first_search(graph)

    print("--- result ---")
    for u in vertices.values():
        print(f"{u.name} ({u.start_time}/{u.finish_time}) : ", end="")
        print_path(u, u.name)
        print("")

File: elementary-graph-algorithms/golang/depth_first_search.go
package main

import "fmt"

var time = 0

func depthFirstSearch(graph *Graph) {
	vertices := graph.vertices
	for _, u := range vertices {
		u.predecessor = nil
		u.color = 0
		u.start_time = 0
		u.end_time = 0
	}

	for _, u := range vertices {
		if u.color == 0 {
			depthFirstSearchInternal(graph, u)
		}
	}
}

func depthFirstSearchInternal(graph *Graph, u *Vertex) {
	time += 1
	fmt.Printf("%s: change color to gray. time = %d\n", u.name, time)
	u.start_time = time
	u.color = 1

	for _, v := range graph.adjacencyList[u.name] {
		if v.color == 0 {
			v.predecessor = u
			depthFirstSearchInternal(graph, v)
		}
	}

	time += 1
	fmt.Printf("%s: change color to black. time = %d\n", u.name, time)
	u.end_time = time
	u.color = 2
}

func printPath(v *Vertex, name string) {
	if v == nil {
		return
	}

	printPath(v.predecessor, name)
	fmt.Printf("%s", v.name)
	if v.name != name {
		fmt.Print(" -> ")
	}
}

func main() {
	vertices := make(map[string]*Vertex, 6)
	for _, name := range "uvxywz" {
		vertices[string(name)] = &Vertex{name: string(name)}
	}

	edges := make([]*Edge, 0, 8)
	edges = append(edges, &Edge{start: "u", end: "v"})
	edges = append(edges, &Edge{start: "u", end: "x"})
	edges = append(edges, &Edge{start: "v", end: "y"})
	edges = append(edges, &Edge{start: "x", end: "v"})
	edges = append(edges, &Edge{start: "y", end: "x"})
	edges = append(edges, &Edge{start: "w", end: "y"})
	edges = append(edges, &Edge{start: "w", end: "z"})
	edges = append(edges, &Edge{start: "z", end: "z"})

	graph := &Graph{vertices: vertices, edges: edges}
	graph.initializeGraph()

	fmt.Println("--- graph ---")
	graph.printGraph()

	fmt.Println("--- depth first search ---")
	depthFirstSearch(graph)

	fmt.Println("--- result ---")
	for _, v := range vertices {
		fmt.Printf("%s (%d/%d): ", v.name, v.start_time, v.end_time)
		printPath(v, v.name)
		fmt.Print("\n")
	}
}
